### PR TITLE
Big refactor

### LINKED ISSUE #16: Not propagating variables correctly
This code:
https://gist.github.com/kress95/cb512facbb1f3077a74daf4e9e15c2ab/66fcb53504ce5904794f728be5d544300a3e5c9a

is yielding this code:
https://gist.github.com/kress95/cb512facbb1f3077a74daf4e9e15c2ab/58d552ccae1fb762ab81293e0bcd541a6be99ff3

that is throwing this error:
https://gist.github.com/kress95/cb512facbb1f3077a74daf4e9e15c2ab/a1afb972713635a9eb1fe5318bb0a1f6af92f6d6

(using latest version on NPM)

### PR DESCRIPTION
**Pull request by Gvozd**
@phpnode, please look my code
I try to create more simple&stable code

Also this refactoring fix #16, #17

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -6,229 +6,125 @@
 const $classConstructor = '__classConstructor';
 const $classMethod = '__classMethod';
 const $objectMethod = '__objectMethod';
+const $bindedArrowFunction = '__bindedArrowFunction';
```
Comment (phpnode) [inline src/index.js]: Let's just use symbols here? Also tiny nitpick - `bound` rather than `binded`

File: src/index.js [hunk 2]
```diff
@@ -6,229 +6,125 @@
...
+      Function: {
+        exit (path) {
+          const {node} = path;
+          if (node[$classConstructor] || node.body[$classConstructor] || node[$classMethod] || node[$objectMethod] || node[$bindedArrowFunction]) {
             return;
           }
-
-          if (bindingInfo && bindingInfo.constant) {
-            state.bindings[node.name] = bindingInfo;
+          if (path.findParent(({node}) => node._generated || node._compact)) {
```
Comment (phpnode) [inline src/index.js]: Let's eliminate this closure 😄

File: src/index.js [hunk 3]
```diff
@@ -6,229 +6,125 @@
...
-      }
     }
+  };
 
-    hasNonParamBindings() {
-      for (let name in this.bindings) {
-        const binding = this.bindings[name];
-        if (binding.kind !== "param") {
-          return true;
+  function getBestHoistedScope(path) {
```
Comment (phpnode) [inline src/index.js]: Perhaps we should use `highest` rather than `best`?

Comment (Gvozd) [inline src/index.js]: Yes, you right

File: src/index.js (hunk 2 above)
Comment (Gvozd) [inline src/index.js]: Maybe better add this plugin to .babelrc? ;)

Comment (phpnode) [inline src/index.js]: Not actually a bad idea, i do that with the typecheck plugin - run it against itself as an extra sanity check.

Comment (Gvozd) [inline src/index.js]: Very nice idea)
I miss polyfill & I found it by self-compiling)

**Pull request merged**